jormungandr/src/network/mod.rs

```diff
@@ -360,61 +337,92 @@ async fn handle_network_input(
     }
 }
 
-async fn handle_propagation_msg(msg: PropagateMsg, state: GlobalStateR, channels: Channels) {
+// propagate message to every peer and return the ones that we could not contact
+async fn propagate_message<F, Fut, E, T>(
+    f: F,
+    sel: poldercast::layer::Selection,
+    arg: T,
+    mbox: &mut MessageBox<TopologyMsg>,
+) -> Result<Vec<topology::Peer>, PropagateError>
+where
+    T: Clone,
+    F: Fn(SocketAddr, T) -> Fut,
+    Fut: Future<Output = Result<(), E>>,
+{
+    let (reply_handle, reply_future) = crate::intercom::unary_reply();
+    mbox.send(TopologyMsg::View(sel, reply_handle)).await?;
+    let peers = reply_future.await.map(|view| view.peers)?;
+
+    // FIXME: this is a workaround because we need to know also the id of the nodes that failed to connect,
+    // it should not be less efficient, just less clean. Remove this once we decided what to do with peers
+    // and ids.
```

Comment:
The way I see this is that until we decide how IDs from topology and peers should interact (and I would rather leave this for when we introduce encryption/verification of peers), passing the topology ID of a peer to the peer propagation function is kind of bleeding/putting together parts that are logically separate, while with this trick we keep everything on its own (logical) plane.

(We need to know the topology ID of a node that failed because then we will quarantine it)

Reply:
For sanity, the same IDs should be used also in our application-level node auth and maybe in a future Noise rework with secure connections (that @NicolasDP proposed).

Reply:
It's something to fix later, though.